eliot-service/eliot/cache_manager.py

```diff
@@ -0,0 +1,366 @@
+        lru_key = str(path)
+        if lru_key in self.lru:
+            size = self.lru.pop(lru_key)
+            self.total_size -= size
+
+    def update_bookkeeping(self, path):
+        """Update bookkeeping for a specified path.
+
+        :arg Path path: the path of the file to update in our bookkeeping
+
+        """
+        lru_key = str(path)
+
+        # If the path is in the cache, remove it
+        if lru_key in self.lru:
+            size = self.lru.pop(lru_key)
+            self.total_size -= size
+            LOGGER.debug(f"removed {path} {size:,d}")
+
+        # If the path exists, calculate new total size, check max size, remove items,
+        # and re-add path.
+        if path.exists():
+            size = path.stat().st_size
+            self.total_size += size
+
+            # If we're beyond the max, then we need to size down as much as we can
+            while self.lru and self.total_size > self.max_size:
+                rm_path, rm_size = self.lru.popitem(last=False)
+                rm_path = pathlib.Path(rm_path)
+                self.total_size -= rm_size
+                rm_path.unlink(missing_ok=True)
+                LOGGER.debug(f"evicted {rm_path} {rm_size:,d}")
+
+            # Since the file exists, add it to the lru
+            LOGGER.debug(f"added {path} {size:,d}")
+            self.lru[lru_key] = size
+
+        LOGGER.debug(f"lru: count {len(self.lru)}, size {self.total_size:,d}")
+
+    def inventory_existing(self):
+        """Walk cache_dir directory and update bookkeeping."""
+        # Reset everything
+        self.total_size = 0
+        self.lru = OrderedDict()
+
+        # Walk directory and update bookkeeping
+        for base, dirs, files in os.walk(str(self.cache_dir)):
+            for fn in files:
+                fn = os.path.join(base, fn)
+                self.update_bookkeeping(pathlib.Path(fn))
+
+    def event_generator(self, nonblocking=False):
+        """Returns a generator of inotify events."""
+        if nonblocking:
+            # NOTE(willkg): Timeout of 0 should return immediately if there's nothing
+            # there
+            timeout = 0
+        else:
+            timeout = 1000
+
+        inotify = inotify_simple.INotify(nonblocking=nonblocking)
+        watch_flags = (
+            flags.ACCESS
+            | flags.CREATE
+            | flags.DELETE
+            | flags.MODIFY
+            | flags.MOVED_FROM
+            | flags.MOVED_TO
+        )
+        watch_descriptor = inotify.add_watch(str(self.cache_dir), watch_flags)
+
+        LOGGER.info("Entering loop")
+        try:
+            while True:
+                for event in inotify.read(timeout=timeout):
+                    LOGGER.debug(event)
+                    event_flags = flags.from_mask(event.mask)
+                    # Ignore changes to directories
+                    if flags.ISDIR in event_flags:
+                        continue
+
+                    path = self.cache_dir / event.name
```

Comment:
The `event.name` is relative to the watched thing which in this case is a directory. If/when we rework this to support a directory tree, we'll have to do a lookup of the watch descriptor in the event back to the path that's being looked at and then use that to generate the full path here.